basket/news/celery.py

```diff
@@ -5,11 +5,44 @@
 # set the default Django settings module for the 'celery' program.
 os.environ.setdefault('DJANGO_SETTINGS_MODULE', 'basket.settings')
 
+from django.conf import settings
+from django.utils.encoding import force_bytes
+
 from celery import Celery as CeleryBase
+from cryptography.fernet import Fernet, MultiFernet, InvalidToken
+from kombu import serialization
+from kombu.utils import json
 from raven.contrib.celery import register_signal, register_logger_signal
 from raven.contrib.django.raven_compat.models import client
 
 
+FERNET = None
+
+if settings.KOMBU_FERNET_KEY:
+    FERNET = Fernet(settings.KOMBU_FERNET_KEY)
+    if settings.KOMBU_FERNET_KEY_PREVIOUS:
+        # this will try both keys. for key rotation.
+        FERNET = MultiFernet([FERNET, Fernet(settings.KOMBU_FERNET_KEY_PREVIOUS)])
+
+
+def fernet_dumps(message):
+    message = json.dumps(message)
+    return FERNET.encrypt(force_bytes(message))
+
+
+def fernet_loads(encoded_message):
+    try:
+        encoded_message = FERNET.decrypt(force_bytes(encoded_message))
+    except InvalidToken:
+        # likely a message that was not encrypted
+        pass
```

Comment:
Is the thinking here that we would be pulling unencrypted messages from the queue right after deploying this? It seems like we might want some sort of notification, such as incrementing a statsd counter, so that we can confirm that there aren’t any unencrypted messages flowing through the system somehow after a certain point.

Reply:
that's a good idea. I've been thinking about this bit and it "shouldn't" ever happen. The way Celery will handle this is that the messages on the queue at deployment won't be encrypted of course, but also they were sent using the normal "json" transport and have a different mime type that is tied to the "json" transport decoder at the other end. So those messages should go through and then stop as the default transport changes at deployment once an encryption key is configured. We may want to just allow the error to happen here so that Sentry will notify us as well as your good suggestion of adding some more instrumentation.